src/nupic/encoders/Encoder.hpp

```diff
@@ -0,0 +1,524 @@
+/* ---------------------------------------------------------------------
+* Numenta Platform for Intelligent Computing (NuPIC)
+* Copyright (C) 2014-2016, Numenta, In  Unless you have an agreement
+* with Numenta, In, for a separate license for this software code, the
+* following terms and conditions apply:
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU Affero Public License version 3 as
+* published by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+* See the GNU Affero Public License for more details.
+*
+* You should have received a copy of the GNU Affero Public License
+* along with this program.  If not, see http://www.gnu.org/licenses.
+*
+* http://numenta.org/licenses/
+* ---------------------------------------------------------------------
+*/
+
+#ifndef NTA_encoder_HPP
+#define NTA_encoder_HPP
+
+//STL includes
+#include <cstring>
+#include <iostream>
+#include <string>
+#include <vector>
+#include <tuple>
+#include <map>
+#include <list>
+#include <cmath>
+
+#include "Utils.hpp" 
+#include "DecodeResult.hpp" 
+#include "RangeList.hpp" 
+
+#include <boost/any.hpp>
+#include <nupic/utils/Log.hpp> 
+
+using namespace std;
+using namespace nupic;
+
+namespace encoders {	 
+
+	// define NAN for VS2012
+#ifndef _HUGE_ENUF
+#define _HUGE_ENUF  1e+300	/* _HUGE_ENUF*_HUGE_ENUF must overflow */
+#endif /* _HUGE_ENUF */
+
+#ifndef INFINITY	
+#define INFINITY   ((float)(_HUGE_ENUF * _HUGE_ENUF))  /* causes warning C4756: overflow in constant arithmetic (by design) */
+#endif /* INFINITY */
+
+#ifndef NAN
+#define NAN        ((float)(INFINITY * 0.0F))
+#endif /* NAN */
+
+	/** Value used to represent no data */
+	static double SENTINEL_VALUE_FOR_MISSING_DATA = NAN;
+
+	template<typename T>										 						
+	T round(T num) {
+		return (num > 0.0) ? floor(num + 0.5) : ceil(num - 0.5);
+	}
+
+
+	/**
+	* <pre>
+	* An encoder takes a value and encodes it with a partial sparse representation
+	* of bits.  The Encoder superclass implements:
```

Comment:
I don't think we need a superclass for encoders:
 * The SDR can do some of this stuff, and is more general purpose.
 * This class contains a lot of implementation specific details which many of the encoders will discard.

Reply:
I think we need a superclass to enforce a basic interface, like: 
```
encode()
encodeIntoArray()
//and basic settings,
```
agree most/all of that class can be removed. Lot if the setters are specific, as you've said.